recipes/LJSpeech/TTS/fastspeech/train.py

```diff
@@ -0,0 +1,404 @@
+            torchaudio.save(path, wav, self.hparams.sample_rate)
+
+def dataio_prepare(hparams):
+    #read saved lexicon
+    with open(os.path.join(hparams["save_folder"], "lexicon"), 'r') as f:
+        lexicon = f.read().split('\t')
+    input_encoder = hparams.get("input_encoder")
+
+    #add a dummy symbol for idx 0 - used for padding.
+    lexicon = ['@@'] + lexicon
+    input_encoder.update_from_iterable(
+                lexicon,
+                sequence_input=False)
+    #load audio, text and durations on the fly; encode audio and text.
+    @sb.utils.data_pipeline.takes("wav", "label", "durations")
+    @sb.utils.data_pipeline.provides("mel_text_pair")
+    def audio_pipeline(wav, label, dur):
+        durs = np.load(dur)
+        durs_seq = torch.from_numpy(durs).int()
+        text_seq = input_encoder.encode_sequence_torch(label.lower()).int()
+        assert len(text_seq) == len(durs) #ensure every token has a duration
+        audio = sb.dataio.dataio.read_audio(wav)
+        mel = hparams["mel_spectogram"](audio=audio)
+        return text_seq, durs_seq, mel, len(text_seq)
+
+    #define splits and load it as sb dataset
+    datasets = {}
+    dataset_names = {
+                    'train': hparams["train_data_path"],
+                    'valid': hparams["valid_data_path"],
+                    'test': hparams["test_data_path"]
+                    }
+
+    for dataset in dataset_names:
+        datasets[dataset] = sb.dataio.dataset.DynamicItemDataset.from_csv(
+            csv_path=os.path.join(hparams["save_folder"], dataset+'.csv'),
+            replacements={"data_root": hparams["data_folder"]},
+            dynamic_items=[audio_pipeline],
+            output_keys=["mel_text_pair", "wav", "label", "durations"],
+        )
+
+    return datasets
+
+
+def batch_to_gpu(batch, return_metadata=False):
+    """Transfers the batch to the target device
+        Arguments
+        ---------
+        batch: tuple
+            the batch to use
+        Returns
+        -------
+        batch: tuple
+            the batch on the correct device
+        """
+
+    (
+        text_padded,
+        durations,
+        input_lengths,
+        mel_padded,
+        output_lengths,
+        len_x,
+        labels,
+        wavs
+    ) = batch
+
+    durations = to_gpu(durations).long()
+    phonemes = to_gpu(text_padded).long()
+    input_lengths = to_gpu(input_lengths).long()
+    max_len = torch.max(input_lengths.data).item()
+    spectogram = to_gpu(mel_padded).float()
+    mel_lengths = to_gpu(output_lengths).long()
+    x = (phonemes, durations)
+    y = (spectogram, durations, mel_lengths, input_lengths)
+    metadata = (labels, wavs)
+    if return_metadata:
+        return x, y, metadata
+    return x, y
+
+def to_gpu(x):
+    x = x.contiguous()
+
+    if torch.cuda.is_available():
+        x = x.cuda(non_blocking=True)
+    return x
+
+
+def criterion(predictions, targets, log_scale_durations=True):
```

Comment:
better to remove the loss func into Fastspeech.py and call it from the yaml file. And for improvement maybe we can try adding L1 for mel_loss too.